## **Aula 1 - Fazendo leitura de arquivos CSV**

Temos um supermercado que está planejando uma liquidação de fim de ano, eles querem oferecer uma assinatura que dá 20% de desconto em todas as compras.

E o time de marketing desse supermercado entende que a melhor maneira de diminuir custos com essa campanha é analisar os dados de compras dos clientes e entender quais teriam mais chances de aderir a essa assinatura e oferecer diretamente a essas pessoas.

Dados: 
- https://github.com/alura-cursos/Pandas/blob/main/superstore_data.csv
- https://github.com/alura-cursos/Pandas/blob/main/superstore_data_ponto_virgula.csv

### **Lendo arquivos CSV**

In [ ]:
import pandas as pd

In [ ]:
url = 'https://raw.githubusercontent.com/alura-cursos/Pandas/main/superstore_data.csv'

In [ ]:
dados = pd.read_csv(url)

In [ ]:
dados.head()

Id  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0   1826        1970  Graduation       Divorced  84835.0        0         0   
1      1        1961  Graduation         Single  57091.0        0         0   
2  10476        1958  Graduation        Married  67267.0        0         1   
3   1386        1967  Graduation       Together  32474.0        1         1   
4   5371        1989  Graduation         Single  21474.0        1         0   

  Dt_Customer  Recency  MntWines  ...  MntFishProducts  MntSweetProducts  \
0   6/16/2014        0       189  ...              111               189   
1   6/15/2014        0       464  ...                7                 0   
2   5/13/2014        0       134  ...               15                 2   
3   11/5/2014        0        10  ...                0                 0   
4    8/4/2014        0         6  ...               11                 0   

   MntGoldProds  NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0           218                  1                4                    4   
1            37                  1                7                    3   
2            30                  1                3                    2   
3             0                  1                1                    0   
4            34                  2                3                    1   

   NumStorePurchases  NumWebVisitsMonth  Response  Complain  
0                  6                  1         1         0  
1                  7                  5         1         0  
2                  5                  2         0         0  
3                  2                  7         0         0  
4                  2                  7         1         0  

[5 rows x 22 columns]

In [ ]:
url_2 = 'https://raw.githubusercontent.com/alura-cursos/Pandas/main/superstore_data_ponto_virgula.csv'

In [ ]:
dados_ponto_virgula = pd.read_csv(url_2)

In [ ]:
dados_ponto_virgula.head()

Id;Year_Birth;Education;Marital_Status;Income;Kidhome;Teenhome;Dt_Customer;Recency;MntWines;MntFruits;MntMeatProducts;MntFishProducts;MntSweetProducts;MntGoldProds;NumDealsPurchases;NumWebPurchases;NumCatalogPurchases;NumStorePurchases;NumWebVisitsMonth;Response;Complain
0  1826;1970;Graduation;Divorced;84835;0;0;6/16/2...                                                                                                                                                                                                                             
1  1;1961;Graduation;Single;57091;0;0;6/15/2014;0...                                                                                                                                                                                                                             
2  10476;1958;Graduation;Married;67267;0;1;5/13/2...                                                                                                                                                                                                                             
3  1386;1967;Graduation;Together;32474;1;1;11/5/2...                                                                                                                                                                                                                             
4  5371;1989;Graduation;Single;21474;1;0;8/4/2014...

### **Parâmetros da função read_csv**
Documentação: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html

In [ ]:
dados_ponto_virgula = pd.read_csv(url_2, sep = ';')

In [ ]:
dados_ponto_virgula

Id  Year_Birth   Education Marital_Status   Income  Kidhome  \
0      1826        1970  Graduation       Divorced  84835.0        0   
1         1        1961  Graduation         Single  57091.0        0   
2     10476        1958  Graduation        Married  67267.0        0   
3      1386        1967  Graduation       Together  32474.0        1   
4      5371        1989  Graduation         Single  21474.0        1   
...     ...         ...         ...            ...      ...      ...   
2235  10142        1976         PhD       Divorced  66476.0        0   
2236   5263        1977    2n Cycle        Married  31056.0        1   
2237     22        1976  Graduation       Divorced  46310.0        1   
2238    528        1978  Graduation        Married  65819.0        0   
2239   4070        1969         PhD        Married  94871.0        0   

      Teenhome Dt_Customer  Recency  MntWines  ...  MntFishProducts  \
0            0   6/16/2014        0       189  ...              111   
1            0   6/15/2014        0       464  ...                7   
2            1   5/13/2014        0       134  ...               15   
3            1   11/5/2014        0        10  ...                0   
4            0    8/4/2014        0         6  ...               11   
...        ...         ...      ...       ...  ...              ...   
2235         1    7/3/2013       99       372  ...               47   
2236         0   1/22/2013       99         5  ...                3   
2237         0   3/12/2012       99       185  ...               15   
2238         0  11/29/2012       99       267  ...              149   
2239         2    1/9/2012       99       169  ...              188   

      MntSweetProducts  MntGoldProds  NumDealsPurchases  NumWebPurchases  \
0                  189           218                  1                4   
1                    0            37                  1                7   
2                    2            30                  1                3   
3                    0             0                  1                1   
4                    0            34                  2                3   
...                ...           ...                ...              ...   
2235                48            78                  2                5   
2236                 8            16                  1                1   
2237                 5            14                  2                6   
2238               165            63                  1                5   
2239                 0           144                  1                8   

      NumCatalogPurchases  NumStorePurchases  NumWebVisitsMonth  Response  \
0                       4                  6                  1         1   
1                       3                  7                  5         1   
2                       2                  5                  2         0   
3                       0                  2                  7         0   
4                       1                  2                  7         1   
...                   ...                ...                ...       ...   
2235                    2                 11                  4         0   
2236                    0                  3                  8         0   
2237                    1                  5                  8         0   
2238                    4                 10                  3         0   
2239                    5                  4                  7         1   

      Complain  
0            0  
1            0  
2            0  
3            0  
4            0  
...        ...  
2235         0  
2236         0  
2237         0  
2238         0  
2239         0  

[2240 rows x 22 columns]

#### **Parâmetro ```nrows```**

Número de linhas a serem lidos do arquivo. Útil para ler pedaços de arquivos grandes.

In [ ]:
dados_primeiras_linhas = pd.read_csv(url,nrows=5)
dados_primeiras_linhas 

Id  Year_Birth   Education Marital_Status  Income  Kidhome  Teenhome  \
0   1826        1970  Graduation       Divorced   84835        0         0   
1      1        1961  Graduation         Single   57091        0         0   
2  10476        1958  Graduation        Married   67267        0         1   
3   1386        1967  Graduation       Together   32474        1         1   
4   5371        1989  Graduation         Single   21474        1         0   

  Dt_Customer  Recency  MntWines  ...  MntFishProducts  MntSweetProducts  \
0   6/16/2014        0       189  ...              111               189   
1   6/15/2014        0       464  ...                7                 0   
2   5/13/2014        0       134  ...               15                 2   
3   11/5/2014        0        10  ...                0                 0   
4    8/4/2014        0         6  ...               11                 0   

   MntGoldProds  NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0           218                  1                4                    4   
1            37                  1                7                    3   
2            30                  1                3                    2   
3             0                  1                1                    0   
4            34                  2                3                    1   

   NumStorePurchases  NumWebVisitsMonth  Response  Complain  
0                  6                  1         1         0  
1                  7                  5         1         0  
2                  5                  2         0         0  
3                  2                  7         0         0  
4                  2                  7         1         0  

[5 rows x 22 columns]

#### **Parâmetro ```usecols```**

Seleciona um subconjunto de colunas a serem lidas. O parâmetro pode receber uma lista de rótulos das colunas ou uma lista de índices

In [ ]:
dados_selecao = pd.read_csv(url, usecols=['Id', 'Year_Birth', 'Income'])
dados_selecao

Id  Year_Birth   Income
0      1826        1970  84835.0
1         1        1961  57091.0
2     10476        1958  67267.0
3      1386        1967  32474.0
4      5371        1989  21474.0
...     ...         ...      ...
2235  10142        1976  66476.0
2236   5263        1977  31056.0
2237     22        1976  46310.0
2238    528        1978  65819.0
2239   4070        1969  94871.0

[2240 rows x 3 columns]

In [ ]:
dados_selecao = pd.read_csv(url, usecols=[0,1,4])
dados_selecao

Id  Year_Birth   Income
0      1826        1970  84835.0
1         1        1961  57091.0
2     10476        1958  67267.0
3      1386        1967  32474.0
4      5371        1989  21474.0
...     ...         ...      ...
2235  10142        1976  66476.0
2236   5263        1977  31056.0
2237     22        1976  46310.0
2238    528        1978  65819.0
2239   4070        1969  94871.0

[2240 rows x 3 columns]

### **Escrevendo arquivos CSV**

Documentação: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html

In [ ]:
dados_selecao.to_csv('clientes_mercado.csv')

In [ ]:
clientes_mercado = pd.read_csv('/content/clientes_mercado.csv')
clientes_mercado

Unnamed: 0     Id  Year_Birth   Income
0              0   1826        1970  84835.0
1              1      1        1961  57091.0
2              2  10476        1958  67267.0
3              3   1386        1967  32474.0
4              4   5371        1989  21474.0
...          ...    ...         ...      ...
2235        2235  10142        1976  66476.0
2236        2236   5263        1977  31056.0
2237        2237     22        1976  46310.0
2238        2238    528        1978  65819.0
2239        2239   4070        1969  94871.0

[2240 rows x 4 columns]

In [ ]:
dados_selecao.to_csv('dados_mercado.csv', index=False)

In [ ]:
dados_mercado = pd.read_csv('/content/dados_mercado.csv')
dados_mercado

Id  Year_Birth   Income
0      1826        1970  84835.0
1         1        1961  57091.0
2     10476        1958  67267.0
3      1386        1967  32474.0
4      5371        1989  21474.0
...     ...         ...      ...
2235  10142        1976  66476.0
2236   5263        1977  31056.0
2237     22        1976  46310.0
2238    528        1978  65819.0
2239   4070        1969  94871.0

[2240 rows x 3 columns]

## **Aula 2 - Utilizando planilhas**

Temos um novo conjunto de dados: um grupo de pesquisa com dados das emissões de gás carbônico (CO2) pelos países ao longo dos anos. Eles querem analisar esses dados com o objetivo de entender os impactos ambientais gerados pelos países e pensar em questões de sustentabilidade. Os dados estão em uma planilha, mas o grupo de pesquisa gostaria de usar a biblioteca Pandas. Será possível?

Para começar, entenderemos o que é uma planilha. De início, vamos abrir a ferramenta Google Planilhas e criar um documento vazio. Na parte superior, teremos colunas nomeadas por letras. Na lateral esquerda, temos colunas numeradas a partir do número um. Nessas planilhas, os dados podem estar nos formato Excel (```.xlsx```) ou CSV

Dados: https://github.com/alura-cursos/Pandas/blob/main/emissoes_CO2.xlsx

### **Lendo arquivos Excel**

No GitHub, temos o arquivo ```emissoes_CO2.xlsx```. Pela extensão .xlsx ao final do nome do arquivo, sabemos que se trata do formato Excel. Ao abri-lo, notaremos que não é possível visualizá-lo no GitHub (como visualizamos o CSV em aulas anteriores)

In [ ]:
import pandas as pd

Nós não conseguimos ter acesso ao dado bruto, pois ele é um arquivo xlsx. Então como leremos o arquivo? Depois de .xlsx, colocaremos um símbolo de interrogação e passar o parâmetro raw=True para trazer o arquivo bruto.

In [ ]:
url = 'https://github.com/alura-cursos/Pandas/blob/main/emissoes_CO2.xlsx?raw=True'

Faremos a leitura do arquivo através da função ```read_excel()``` da biblioteca Pandas

In [ ]:
dados_co2 = pd.read_excel(url)

A execução pode demorar um pouco, porque o arquivo é grande. Na sequência, vamos consultar as cinco primeiras linhas do dataframe gerado, com a função ```head()```

In [ ]:
dados_co2.head()

País ISO 3166-1 alpha-3     Ano  Total  Carvão  Oléo  Gás  Cimento  \
0  Afeganistão                AFG  1750.0    0.0     NaN   NaN  NaN      NaN   
1  Afeganistão                AFG  1751.0    0.0     NaN   NaN  NaN      NaN   
2  Afeganistão                AFG  1752.0    0.0     NaN   NaN  NaN      NaN   
3  Afeganistão                AFG  1753.0    0.0     NaN   NaN  NaN      NaN   
4  Afeganistão                AFG  1754.0    0.0     NaN   NaN  NaN      NaN   

   Queima  Outro  Per Capita  
0     NaN    NaN         NaN  
1     NaN    NaN         NaN  
2     NaN    NaN         NaN  
3     NaN    NaN         NaN  
4     NaN    NaN         NaN

Como comentamos, uma planilha pode ter diferentes páginas. Para descobrir se há mais páginas, usaremos uma classe da biblioteca Pandas chamada ```ExcelFile()```. Entre os parênteses, passaremos a URL. Depois, empregamos o atributo ```sheet_names``` (nome das páginas):

In [ ]:
pd.ExcelFile(url).sheet_names

['emissoes_C02', 'emissoes_percapita', 'fontes']

A execução pode ser um pouco demorada. Como resultado, temos os nomes das três páginas. Para especificar uma página (entre outras ações), podemos explorar alguns parâmetros da função ```read_excel()```

### **Parâmetros da função read_excel**

Documentação: https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html

#### Parâmetro sheet_name
O parâmetro ```sheet_name``` pode ser utilizado para especificar a página da planilha que queremos mostrar.

Por padrão, esse parâmetro é igual a zero. Ou seja, quando não alteramos esse valor, a primeira página será mostrada. Já se definirmos ```sheet_name=1```, consultaremos a segunda página. Alternativamente, podemos escrever o nome da página, por exemplo, ```sheet_name="Pagina2"```.

In [ ]:
percapita = pd.read_excel(url, sheet_name='emissoes_percapita')

In [ ]:
percapita.head()

País ISO 3166-1 alpha-3     Ano  Total  Carvão  Óleo  Gás  Cimento  \
0  Afeganistão                AFG  1750.0    NaN     NaN   NaN  NaN      NaN   
1  Afeganistão                AFG  1751.0    NaN     NaN   NaN  NaN      NaN   
2  Afeganistão                AFG  1752.0    NaN     NaN   NaN  NaN      NaN   
3  Afeganistão                AFG  1753.0    NaN     NaN   NaN  NaN      NaN   
4  Afeganistão                AFG  1754.0    NaN     NaN   NaN  NaN      NaN   

   Queima  Outro  
0     NaN    NaN  
1     NaN    NaN  
2     NaN    NaN  
3     NaN    NaN  
4     NaN    NaN

In [ ]:
fontes = pd.read_excel(url, sheet_name='fontes')

In [ ]:
fontes.head()

País ISO 3166-1 alpha-3     Ano   Total  Carvão    Oléo     Gás  \
0  Afeganistão                AFG  1750.0  [NONE]  [NONE]  [NONE]  [NONE]   
1  Afeganistão                AFG  1751.0  [NONE]  [NONE]  [NONE]  [NONE]   
2  Afeganistão                AFG  1752.0  [NONE]  [NONE]  [NONE]  [NONE]   
3  Afeganistão                AFG  1753.0  [NONE]  [NONE]  [NONE]  [NONE]   
4  Afeganistão                AFG  1754.0  [NONE]  [NONE]  [NONE]  [NONE]   

  Cimento  Queima   Outro Per Capita  
0  [NONE]  [NONE]  [NONE]     [NONE]  
1  [NONE]  [NONE]  [NONE]     [NONE]  
2  [NONE]  [NONE]  [NONE]     [NONE]  
3  [NONE]  [NONE]  [NONE]     [NONE]  
4  [NONE]  [NONE]  [NONE]     [NONE]

#### Parâmetro usecols

Especifica as colunas que serão lidas.

Como aprendemos anteriormente, nas planilhas as colunas são nomeadas por letras. Ao indicar A:D, selecionamos o intervalo da coluna A até a coluna D

In [ ]:
intervalo = pd.read_excel(url, sheet_name='emissoes_C02', usecols= 'A:D')

In [ ]:
intervalo

País ISO 3166-1 alpha-3     Ano         Total
0      Afeganistão                AFG  1750.0  0.000000e+00
1      Afeganistão                AFG  1751.0  0.000000e+00
2      Afeganistão                AFG  1752.0  0.000000e+00
3      Afeganistão                AFG  1753.0  0.000000e+00
4      Afeganistão                AFG  1754.0  0.000000e+00
...            ...                ...     ...           ...
63099       Global                WLD  2017.0  3.609674e+10
63100       Global                WLD  2018.0  3.682651e+10
63101       Global                WLD  2019.0  3.708256e+10
63102       Global                WLD  2020.0  3.526409e+10
63103       Global                WLD  2021.0  3.712385e+10

[63104 rows x 4 columns]

Agora, temos um dataframe com 63104 linhas (rows) e apenas 4 colunas (coluns). Ou seja, selecionamos apenas o intervalo com nossas colunas de interesse: o nome do país, o código do país, o ano e o valor total de emissão de gás carbônico.

Definir as colunas é especialmente útil quando temos uma página com mais de uma tabela. Assim, conseguimos selecionar somente o intervalo desejado.

#### Parâmetro nrows

Outro parâmetro da função read_excel() é o nrows, responsável por definir a quantidade a linhas exibidas.

In [ ]:
intervalo_2 = pd.read_excel(url, sheet_name='emissoes_C02', usecols= 'A:D', nrows=10)

In [ ]:
intervalo_2

País ISO 3166-1 alpha-3     Ano  Total
0  Afeganistão                AFG  1750.0    0.0
1  Afeganistão                AFG  1751.0    0.0
2  Afeganistão                AFG  1752.0    0.0
3  Afeganistão                AFG  1753.0    0.0
4  Afeganistão                AFG  1754.0    0.0
5  Afeganistão                AFG  1755.0    0.0
6  Afeganistão                AFG  1756.0    0.0
7  Afeganistão                AFG  1757.0    0.0
8  Afeganistão                AFG  1758.0    0.0
9  Afeganistão                AFG  1759.0    0.0

### **Escrevendo arquivos Excel**

Anteriormente, consultamos as cinco primeiras linhas do dataframe com os dados per capita com o seguinte comando:

Nosso grupo de pesquisa deseja analisar somente os valores per capita, então vamos salvá-los esse arquivo em formato Excel para mantê-los separados.

Vamos usar o nome do dataframe "percapita" e usar a função ```to_excel()``` para realizar a escrita no formato Excel. Dentro dos parênteses, passaremos dois parâmetros. O primeiro será o nome do arquivo que queremos salvar, no caso, ```co2_percapita.xlsx```. O segundo parâmetro será o index, que estudamos anteriormente. No caso, o deixaremos com valor ```False```

In [ ]:
percapita.to_excel('co2_percapita.xlsx', index=False)

In [ ]:
pd.read_excel('/content/co2_percapita.xlsx')

País ISO 3166-1 alpha-3   Ano      Total     Carvão       Óleo  \
0      Afeganistão                AFG  1750        NaN        NaN        NaN   
1      Afeganistão                AFG  1751        NaN        NaN        NaN   
2      Afeganistão                AFG  1752        NaN        NaN        NaN   
3      Afeganistão                AFG  1753        NaN        NaN        NaN   
4      Afeganistão                AFG  1754        NaN        NaN        NaN   
...            ...                ...   ...        ...        ...        ...   
63099       Global                WLD  2017  4749682.0  1908857.0  1610910.0   
63100       Global                WLD  2018  4792753.0  1919213.0  1596350.0   
63101       Global                WLD  2019  4775633.0  1896468.0  1589920.0   
63102       Global                WLD  2020  4497423.0  1807760.0  1427353.0   
63103       Global                WLD  2021  4693699.0  1893923.0  1496614.0   

             Gás   Cimento   Queima    Outro  
0            NaN       NaN      NaN      NaN  
1            NaN       NaN      NaN      NaN  
2            NaN       NaN      NaN      NaN  
3            NaN       NaN      NaN      NaN  
4            NaN       NaN      NaN      NaN  
...          ...       ...      ...      ...  
63099   940144.0  198416.0  51579.0  39776.0  
63100   979965.0  204225.0  53634.0  39366.0  
63101   984878.0  208309.0  56569.0  39490.0  
63102   963695.0  208844.0  51981.0  37789.0  
63103  1001585.0  211472.0  52663.0  37443.0  

[63104 rows x 10 columns]

### **Lendo dados do Google Planilhas**

No Google Planilhas, é possível carregar dados no formato Excel e CSV. Uma vez carregados, vamos visualizar as páginas dessa planilha — algo que não tínhamos conseguido fazer no GitHub.

No Google Planilhas, é possível obter o link da planilha e compartilhá-lo com outras pessoas. Além disso, podemos usar esse link para acessar a planilha no Colab!

**Link da planilha**: https://docs.google.com/spreadsheets/d/1lzq0k-41-MbbS63C3Q9i1wPvLkSJt9zhr4Jolt1vEog/edit?usp=sharing

Toda planilha possui um ID para identificá-la. Este ID está presente nessa URL, entre ```/d/``` e ```/edit?usp=sharing```. No caso, o ID da planilha é o seguinte: ```1lzq0k-41-MbbS63C3Q9i1wPvLkSJt9zhr4Jolt1vEog```

Para acessar a planilha no Colab, precisaremos alterar alguns trechos da URL. Podemos utilizar uma f-string para formatar strings e adaptar a nossa url.

Todos os elementos que aparecem entre chaves serão substituídos pelas variáveis correspondentes. No caso, o trecho ```{sheet_id}``` da URL será substituído pela string que armazenamos na variável ```sheet_id``` anteriormente. Essa prática torna o código mais simples, melhorando sua escrita e a leitura.

A seguir, adicionaremos alguns parâmetros para permitir o acesso à planilha. Vamos colocar uma barra após ```{sheet_id}``` e inserir o parâmetro ```gviz/tq```.

Para ter acesso à planilha, é necessário acessar a API de visualização de dados do Google. API significa Application Programming Interface, isto é, uma interface de programação de aplicações. Elas permitem sistemas diferentes se comuniquem.
No caso, queremos criar a comunicação entre o Google Colab e o Google Planilhas. O parâmetro ```gviz/tq``` permite o acesso à API responsável por essa comunicação.

Após ```gviz/tq```, adicionaremos um ponto de interrogação e o parâmetro ```tqx=out:csv```. Dessa forma, nossos dados serão retornados em formato CSV (que já temos familiaridade, de aulas anteriores). Por fim, incluiremos o símbolo do "e comercial" e a palavra ```sheet```. Assim, todos os dados da planilha serão retornados, incluindo cabeçalho, linhas e colunas. Nossa URL está pronta!

In [ ]:
sheet_id = '1lzq0k-41-MbbS63C3Q9i1wPvLkSJt9zhr4Jolt1vEog'
url = f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet'

In [ ]:
dados_co2_sheets = pd.read_csv(url)

In [ ]:
dados_co2_sheets

País ISO 3166-1 alpha-3   Ano           Total          Carvão  \
0      Afeganistão                AFG  1750               0             NaN   
1      Afeganistão                AFG  1751               0             NaN   
2      Afeganistão                AFG  1752               0             NaN   
3      Afeganistão                AFG  1753               0             NaN   
4      Afeganistão                AFG  1754               0             NaN   
...            ...                ...   ...             ...             ...   
63099       Global                WLD  2017  36.096.739.276  14.506.973.805   
63100       Global                WLD  2018  36.826.506.600  14.746.830.688   
63101       Global                WLD  2019  37.082.558.969  14.725.978.025   
63102       Global                WLD  2020  35.264.085.734  14.174.564.010   
63103       Global                WLD  2021  37.123.850.352  14.979.598.083   

                 Oléo            Gás        Cimento       Queima        Outro  \
0                 NaN            NaN            NaN          NaN          NaN   
1                 NaN            NaN            NaN          NaN          NaN   
2                 NaN            NaN            NaN          NaN          NaN   
3                 NaN            NaN            NaN          NaN          NaN   
4                 NaN            NaN            NaN          NaN          NaN   
...               ...            ...            ...          ...          ...   
63099  12.242.627.935  7.144.928.128  1.507.923.185  391.992.176  302.294.047   
63100  12.266.016.285  7.529.846.784  1.569.218.392  412.115.746  302.478.706   
63101  12.345.653.374  7.647.528.220  1.617.506.786  439.253.991  306.638.573   
63102  11.191.808.551  7.556.290.283  1.637.537.532  407.583.673  296.301.685   
63103  11.837.159.116  7.921.829.472  1.672.592.372  416.525.563  296.145.746   

      Per Capita  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
63099  4.749.682  
63100  4.792.753  
63101  4.775.633  
63102  4.497.423  
63103  4.693.699  

[63104 rows x 11 columns]

Retornando outras páginas
Verificamos anteriormente que já três páginas diferentes em nossa planilha, recapitulando: emissoes_CO2, emissoes_percapita, fontes

Para especificar a página da planilha que queremos acessar, basta inserir um símbolo de igual ao final da URL, seguida do nome da página que queremos ler. Vamos armazenar o nome da página na variável ```sheet_name``` e utilizá-la na f-string

In [ ]:
sheet_id = '1lzq0k-41-MbbS63C3Q9i1wPvLkSJt9zhr4Jolt1vEog'
sheet_name = 'emissoes_percapita'
url_percapita = f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={sheet_name}'

In [ ]:
percapita_sheets = pd.read_csv(url_percapita)

In [ ]:
percapita_sheets.head()

País ISO 3166-1 alpha-3   Ano Total Carvão Óleo  Gás Cimento Queima  \
0  Afeganistão                AFG  1750   NaN    NaN  NaN  NaN     NaN    NaN   
1  Afeganistão                AFG  1751   NaN    NaN  NaN  NaN     NaN    NaN   
2  Afeganistão                AFG  1752   NaN    NaN  NaN  NaN     NaN    NaN   
3  Afeganistão                AFG  1753   NaN    NaN  NaN  NaN     NaN    NaN   
4  Afeganistão                AFG  1754   NaN    NaN  NaN  NaN     NaN    NaN   

   Outro  
0    NaN  
1    NaN  
2    NaN  
3    NaN  
4    NaN

In [ ]:
sheet_id = '1lzq0k-41-MbbS63C3Q9i1wPvLkSJt9zhr4Jolt1vEog'
sheet_name = 'fontes'
url_fontes = f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={sheet_name}'

In [ ]:
fontes_sheets = pd.read_csv(url_fontes)

In [ ]:
fontes_sheets.head()

País ISO 3166-1 alpha-3   Ano   Total  Carvão    Oléo     Gás  \
0  Afeganistão                AFG  1750  [NONE]  [NONE]  [NONE]  [NONE]   
1  Afeganistão                AFG  1751  [NONE]  [NONE]  [NONE]  [NONE]   
2  Afeganistão                AFG  1752  [NONE]  [NONE]  [NONE]  [NONE]   
3  Afeganistão                AFG  1753  [NONE]  [NONE]  [NONE]  [NONE]   
4  Afeganistão                AFG  1754  [NONE]  [NONE]  [NONE]  [NONE]   

  Cimento  Queima   Outro Per Capita  
0  [NONE]  [NONE]  [NONE]     [NONE]  
1  [NONE]  [NONE]  [NONE]     [NONE]  
2  [NONE]  [NONE]  [NONE]     [NONE]  
3  [NONE]  [NONE]  [NONE]     [NONE]  
4  [NONE]  [NONE]  [NONE]     [NONE]

## **Aula 3 - Manipulando arquivos JSON**

Agora, temos um hospital que deseja acessar os dados de pacientes, mas os arquivos estão todos no formato JSON. Dessa forma, é difícil extrair informações relevantes a partir dos dados. O nosso objetivo será usar a biblioteca Pandas para ler e manipular esses arquivos.

Dados: 
- https://github.com/alura-cursos/Pandas/blob/main/pacientes.json
- https://github.com/alura-cursos/Pandas/blob/main/pacientes_2.json

### O que é um arquivo JSON?

A seguir, temos um exemplo de informações de um arquivo JSON com dados de uma paciente do hospital

```javascript
{"ID":"01", "Faixa_etaria":"55-59", "Sexo_biologico":"Feminino"}
```

Neste arquivo, temos três informações a respeito da pessoa:
* Um ID atribuído a ela ("ID");
* Sua faixa etária ("Faixa_etaria");
* Seu sexo biológico ("Sexo_biologico").

Estas informações são o que chamamos de chaves. Uma chave é o que indexa, ou seja, posiciona um elemento dentro do arquivo JSON. Cada chave é acompanhada de seus respectivos valores.

Desse modo, o ID da paciente carrega o valor ```"01"```, enquanto a sua faixa etária corresponde à faixa entre 55 e 59 anos (```"55-59"```) e o sexo biológico foi atribuído como ```"Feminino"```.

Os valores de cada chave podem ser números do tipo inteiro (```int```) ou ```float```, podem ser caracteres de texto (```string```) ou uma lista com vários elementos.

 Sempre haverá o símbolo de abrir e fechar chaves no começo e no fim das informações. Nesse sentido, o arquivo se assemelha muito a um dicionário em Python.

 ### O que significa JSON?

 JSON é uma sigla em inglês para JavaScript Object Notation. Em português, são os Objetos de Notação JavaScript. Este é um formato de texto leve baseado em um subconjunto da linguagem de programação JavaScript. Ele é muito utilizado em APIs também.

### **Lendo arquivos JSON**

Então, um arquivo JSON (Javascript Object Notation) é um conjunto de pares chaves / valor.

In [ ]:
import pandas as pd

A função responsável por fazer a leitura dos arquivos em formato JSON é a ```pd.read_json()```.

In [ ]:
dados_pacientes = pd.read_json('/content/pacientes.json')

In [ ]:
dados_pacientes

ID_paciente Doenca_cardiaca    IMC Fumante Consumo_alcool  AVC  \
0              0             Nao  16.60     Sim            Nao  Nao   
1              1             Nao  20.34     Nao            Nao  Sim   
2              2             Nao  26.58     Sim            Nao  Nao   
3              3             Nao  24.21     Nao            Nao  Nao   
4              4             Nao  23.71     Nao            Nao  Nao   
..           ...             ...    ...     ...            ...  ...   
995          995             Nao  28.29     Sim            Nao  Nao   
996          996             Nao  25.69     Sim            Nao  Nao   
997          997             Sim  26.99     Nao            Nao  Nao   
998          998             Nao  44.29     Nao            Nao  Nao   
999          999             Nao  25.83     Nao            Nao  Nao   

     Saude_fisica  Saude_mental Dificuldade_caminhar Sexo_biologico  \
0               3            30                  Nao       Feminino   
1               0             0                  Nao       Feminino   
2              20            30                  Nao      Masculino   
3               0             0                  Nao       Feminino   
4              28             0                  Sim       Feminino   
..            ...           ...                  ...            ...   
995            10            20                  Nao       Feminino   
996             0             0                  Nao       Feminino   
997             1             0                  Nao      Masculino   
998            30             0                  Sim       Feminino   
999             0             0                  Nao      Masculino   

    Faixa_etaria    Raca Diabetes Atividade_fisica Saude_geral  Horas_sono  \
0          55-59  Branca      Sim              Sim   Muito boa           5   
1        80 ou +  Branca      Nao              Sim   Muito boa           7   
2          65-69  Branca      Sim              Sim    Razoavel           8   
3          75-79  Branca      Nao               No         Boa           6   
4          40-44  Branca      Nao              Sim   Muito boa           8   
..           ...     ...      ...              ...         ...         ...   
995        65-69   Preta      Sim              Sim    Razoavel           8   
996        75-79  Branca      Nao              Sim   Excelente           8   
997        70-74  Branca      Nao              Sim   Muito boa           8   
998        70-74  Branca      Sim               No         Boa           6   
999        70-74   Preta      Sim               No   Excelente           7   

    Asma Doenca_renal Cancer_pele  
0    Sim          Nao         Sim  
1    Nao          Nao         Nao  
2    Sim          Nao         Nao  
3    Nao          Nao         Sim  
4    Nao          Nao         Nao  
..   ...          ...         ...  
995  Nao          Nao         Nao  
996  Nao          Nao         Nao  
997  Nao          Nao         Nao  
998  Sim          Nao         Nao  
999  Nao          Nao         Nao  

[1000 rows x 19 columns]

Deu tudo certo com o arquivo de pacientes, mas nem sempre é tão simples fazer a leitura de um arquivo JSON. Algumas vezes, as informações podem estar aninhadas dentro do arquivo JSON. Mas o que seriam informações aninhadas? Para entendermos isso melhor, carregaremos outro arquivo.

In [ ]:
dados_pacientes_2 = pd.read_json('/content/pacientes_2.json')

In [ ]:
dados_pacientes_2

Pesquisa   Ano  \
0  Principais Indicadores de Doenca Cardiaca  2020   
1  Principais Indicadores de Doenca Cardiaca  2020   
2  Principais Indicadores de Doenca Cardiaca  2020   

                                           Pacientes  
0  {'ID': '01', 'Faixa_etaria': '55-59', 'Sexo_bi...  
1  {'ID': '02', 'Faixa_etaria': '80 ou +', 'Sexo_...  
2  {'ID': '03', 'Faixa_etaria': '65-69', 'Sexo_bi...

O resultado é uma tabela com três colunas. A primeira coluna, intitulada "Pesquisa", tem as mesmas informações em todas as linhas, ou seja, o texto "Principais Indicadores de Doença Cardíaca". A segunda coluna exibe o ano em que a pesquisa foi feita e a última coluna, chamada "Pacientes", agrupou todas as informações de cada paciente em uma única célula.

Com isso, não conseguimos obter a separação de cada dado, tampouco ler todas as informações disponíveis, uma vez que o conteúdo da célula está cortado (conseguimos ver o ID e a faixa etária).

Este é um caso típico de informações aninhadas. Não há separação das colunas e das linhas de maneira a facilitar a leitura e a manipulação dos dados.

### **Normalizando arquivos JSON**

Para normalizar esses dados em formato JSON utilizaremos a ```json_normalize()```. Como parâmetro, passaremos o data frame que desejamos normalizar ```dados_pacientes_2``` e indicando qual coluna deve ser normalizada, ```Pacientes```

In [ ]:
df_normalizado = pd.json_normalize(dados_pacientes_2['Pacientes'])

In [ ]:
df_normalizado

ID Faixa_etaria Sexo_biologico    Raça    IMC Fumante Consumo_alcool  \
0  01        55-59       Feminino  Branca  16.60     Sim            Nao   
1  02      80 ou +       Feminino  Branca  20.34     Nao            Nao   
2  03        65-69      Masculino  Branca  26.58     Sim            Nao   

   Saude_fisica  Saude_mental Dificuldade_caminhar Atividade_fisica  \
0             3            30                  Nao              Sim   
1             0             0                  Nao              Sim   
2            20            30                  Nao              Sim   

  Saude_geral  Horas_sono                Problemas_saude  
0   Muito boa           5  [Diabetes, Asma, Cancer_pele]  
1   Muito boa           7                          [AVC]  
2   Muito boa           8               [diabetes, Asma]

Agora, conseguimos obter o data frame normalizado, com uma coluna para cada informação. É isso que a função json_normalize faz: transforma tudo o que estiver aninhado em uma única célula em diferentes linhas e colunas.

### **Escrevendo arquivos JSON**

Para salvar nosso arquivo normalizado, basta utilizar a função ```.to_json()```

In [ ]:
df_normalizado.to_json('historico_pacientes_norm.json')

In [ ]:
pd.read_json('/content/historico_pacientes_norm.json')

ID Faixa_etaria Sexo_biologico    Raça    IMC Fumante Consumo_alcool  \
0   1        55-59       Feminino  Branca  16.60     Sim            Nao   
1   2      80 ou +       Feminino  Branca  20.34     Nao            Nao   
2   3        65-69      Masculino  Branca  26.58     Sim            Nao   

   Saude_fisica  Saude_mental Dificuldade_caminhar Atividade_fisica  \
0             3            30                  Nao              Sim   
1             0             0                  Nao              Sim   
2            20            30                  Nao              Sim   

  Saude_geral  Horas_sono                Problemas_saude  
0   Muito boa           5  [Diabetes, Asma, Cancer_pele]  
1   Muito boa           7                          [AVC]  
2   Muito boa           8               [diabetes, Asma]

## **Aula 4 - Lendo dados em HTML e XML**

Uma pessoa está criando um sistema de recomendação de filmes. Mas, para isso, ela precisa de dados de filmes, de avaliações que as pessoas deram para eles.

Ela encontrou um [artigo web sobre os 100 melhores filmes listados pelo Instituto Americano de Cinema](https://en.wikipedia.org/wiki/AFI%27s_100_Years...100_Movies). Esse artigo está na Wikipedia e, na lateral direita da página, temos uma tabela contendo outros artigos criados pelo mesmo instituto, de 1998 a 2008. Há uma versão atualizada de 2007 do nosso artigo em questão, mas verificaremos a de 1988 primeiro.

Descendo a página, encontramos uma tabela interessante para o projeto, contendo as seguintes colunas: nome do filme, ano de lançamento, diretor, produtora e a posição no ranking, indo do primeiro ao centésimo lugar.

Por fim, mais abaixo na página, encontramos um quadro que parece uma tabela contendo apenas os nomes dos filmes, sua posição no ranking e seus respectivos links para a página da Wikipedia, onde podemos checar a sinopse e conhecer mais sobre eles.

E agora, como podemos carregar esses dados utilizando a biblioteca Pandas?

Para começar, vamos salvar essa página em HTML


Dados: https://github.com/alura-cursos/Pandas/blob/main/filmes_wikipedia.html

### **Lendo páginas web**

In [ ]:
import pandas as pd

Vamos utilizar a função de leitura de arquivos HTML ```.read_html()```

In [ ]:
dados_html = pd.read_html('/content/filmes_wikipedia.html')

In [ ]:
dados_html

[                                                    0  \
 0                                                1998   
 1                                                1999   
 2                                                2000   
 3                                                2001   
 4                                                2002   
 5                                                2003   
 6                                                2004   
 7                                                2005   
 8                                                2005   
 9                                                2006   
 10                                               2006   
 11                                               2007   
 12                                               2008   
 13  .mw-parser-output .hlist dl,.mw-parser-output ...   
 
                                                     1  
 0                                          100 Movies  
 1            

Ao executar essa célula, não recebemos um dataframe. Aparentemente é uma lista, porque começa com colchetes. Primeiro temos os dados daquela primeira tabela, começando em 1998 e indo até 2008. Abaixo disso, temos os dados da segunda tabela e, por fim, os dados da terceira.

***Mas, por que não surgiu um dataframe?***

A função ```read_html()``` busca pelas tags ```<table>``` e transforma esses dataframes em uma lista. Ou seja, nossa lista contém os dataframes um abaixo do outro. Podemos confirmar isso ao verificar o tipo da variável criada, ```dados_html```

In [ ]:
type(dados_html)

list

In [ ]:
len(dados_html)

3

### Acessando um dataframe específico
Se quisermos acessar um dataframe específico entre esses três, podemos simplesmente usar sua posição na lista

In [ ]:
top_filmes = dados_html[1]

In [ ]:
top_filmes

Film  Release year              Director  \
0                   Citizen Kane          1941          Orson Welles   
1                     Casablanca          1942        Michael Curtiz   
2                  The Godfather          1972  Francis Ford Coppola   
3             Gone with the Wind          1939        Victor Fleming   
4             Lawrence of Arabia          1962            David Lean   
..                           ...           ...                   ...   
95                 The Searchers          1956             John Ford   
96              Bringing Up Baby          1938          Howard Hawks   
97                    Unforgiven          1992        Clint Eastwood   
98  Guess Who's Coming to Dinner          1967        Stanley Kramer   
99           Yankee Doodle Dandy          1942        Michael Curtiz   

                      Production companies  Rank  
0                       RKO Radio Pictures     1  
1                    Warner Bros. Pictures     2  
2   Paramount Pictures, Alfran Productions     3  
3          Selznick International Pictures     4  
4                         Horizon Pictures     5  
..                                     ...   ...  
95                  C. V. Whitney Pictures    96  
96                      RKO Radio Pictures    97  
97                     The Malpaso Company    98  
98                       Columbia Pictures    99  
99                   Warner Bros. Pictures   100  

[100 rows x 5 columns]

### **Escrevendo arquivos HTML**

Escrever um arquivo no formato HTML pode ser especialmente útil quando você deseja compartilhar os seus dados com pessoas que não têm acesso à biblioteca Pandas ou outras ferramentas de carregamento de dados.

Os dados podem ser consultados diretamente no navegador ou incorporados numa página web, facilitando muito a visualização.

A função ```.to_html()``` é responsável por escrever esses dados em um arquivo html.

In [ ]:
top_filmes.to_html('top_filmes.html')

In [ ]:
pd.read_html('/content/top_filmes.html')

[    Unnamed: 0                          Film  Release year  \
 0            0                  Citizen Kane          1941   
 1            1                    Casablanca          1942   
 2            2                 The Godfather          1972   
 3            3            Gone with the Wind          1939   
 4            4            Lawrence of Arabia          1962   
 ..         ...                           ...           ...   
 95          95                 The Searchers          1956   
 96          96              Bringing Up Baby          1938   
 97          97                    Unforgiven          1992   
 98          98  Guess Who's Coming to Dinner          1967   
 99          99           Yankee Doodle Dandy          1942   
 
                 Director                    Production companies  Rank  
 0           Orson Welles                      RKO Radio Pictures     1  
 1         Michael Curtiz                   Warner Bros. Pictures     2  
 2   Francis Ford Co

Conseguimos pegar aquela tabela e disponibilizá-la num formato possível de abrir no navegador e também incorporar numa página web. Você pode, inclusive, melhorar essa visualização se tiver alguma familiaridade com o front-end.

Além do formato HTML, podemos salvar esses dados em outros formatos. A pessoa que está criando esse sistema de recomendação provavelmente vai querer tratar esses dados e adicionar informações novas. Para isso, é interessante ter um arquivo no formato CSV ou um formato Excel para trabalhar mais facilmente.

In [ ]:
top_filmes.to_csv('top_filmes_1998', index=False)

In [ ]:
dados = pd.read_csv('/content/top_filmes_1998')
dados.head()

Film  Release year              Director  \
0        Citizen Kane          1941          Orson Welles   
1          Casablanca          1942        Michael Curtiz   
2       The Godfather          1972  Francis Ford Coppola   
3  Gone with the Wind          1939        Victor Fleming   
4  Lawrence of Arabia          1962            David Lean   

                     Production companies  Rank  
0                      RKO Radio Pictures     1  
1                   Warner Bros. Pictures     2  
2  Paramount Pictures, Alfran Productions     3  
3         Selznick International Pictures     4  
4                        Horizon Pictures     5

### **Lendo/escrevendo arquivos XML**

Agora temos um arquivo mais detalhado sobre os filmes, porém ele está no formato XML. Ele é o arquivo imdb_top_1000.xml; ou seja, os dados desse arquivo foram obtidos do IMDB, uma base de dados que contém informações sobre filmes e séries, onde as pessoas atribuem uma nota às obras que assistem.

Dados: https://github.com/alura-cursos/Pandas/blob/main/imdb_top_1000.xml

#### O que é XML?
Esse nome é uma sigla da expressão em inglês Extensible Markup Language, ou "Linguagem de marcação estendida". Sua estrutura de tags é muito parecida com o HTML.

No nosso arquivo, temos uma tag logo no início chamada ```<data>```. Essa é a tag raiz que indica o início do documento. Em seguida temos uma tag chamada ```<row>```. "linha" em português. Para cada linha, temos várias tags que contêm as informações que estarão em cada uma das colunas.

In [ ]:
import pandas as pd

A função ```pd.read_xml()``` é a responsável por realizar a leitura de um arquivo XML

In [ ]:
dados_imdb = pd.read_xml('imdb_top_1000.xml')

In [ ]:
dados_imdb.head(3)

index                                        Poster_Link  \
0      0  https://m.media-amazon.com/images/M/MV5BMDFkYT...   
1      1  https://m.media-amazon.com/images/M/MV5BM2MyNj...   
2      2  https://m.media-amazon.com/images/M/MV5BMTMxNT...   

               Series_Title Released_Year Certificate  Runtime  \
0  The Shawshank Redemption          1994           A  142 min   
1             The Godfather          1972           A  175 min   
2           The Dark Knight          2008          UA  152 min   

                  Genre  IMDB_Rating  \
0                 Drama          9.3   
1          Crime, Drama          9.2   
2  Action, Crime, Drama          9.0   

                                            Overview  Meta_score  \
0  Two imprisoned men bond over a number of years...        80.0   
1  An organized crime dynasty's aging patriarch t...       100.0   
2  When the menace known as the Joker wreaks havo...        84.0   

               Director           Star1           Star2          Star3  \
0        Frank Darabont     Tim Robbins  Morgan Freeman     Bob Gunton   
1  Francis Ford Coppola   Marlon Brando       Al Pacino     James Caan   
2     Christopher Nolan  Christian Bale    Heath Ledger  Aaron Eckhart   

            Star4  No_of_Votes        Gross  
0  William Sadler      2343110   28,341,469  
1    Diane Keaton      1620367  134,966,411  
2   Michael Caine      2303232  534,858,444

A função ```to_xml()``` é a responsável por realizar a escrita de um arquivo XML. Essa função, para salvar no formato XML, é muito útil quando trabalhamos com bancos de dados ou outros sistemas que se integrem facilmente a esse formato

In [ ]:
dados_imdb.to_xml('filmes_imdb.xml')